scalameta/tokenizers/shared/src/main/scala/scala/meta/internal/tokenizers/XmlParser.scala

```diff
@@ -11,124 +11,114 @@ import fastparse.all._
-  val ContentP1 = P(ElemPattern | Reference | CDSect | PI | Comment | ScalaPatterns)
-  val XmlExpr = P(WL ~ XmlContent ~ (WL ~ Element).rep)
-  val XmlPattern = P(WL ~ ElemPattern)
+                WL: P0 = CharsWhile(_.isWhitespace).opaque("whitespace")) {
+
+  val XmlExpr: P0 = P( Xml.XmlContent.rep(min = 1, sep = WL.?) )
+  val XmlPattern: P0 = P( Xml.ElemPattern )
+
+  private[this] object Xml {
+    val Element   = P( TagHeader ~/ ("/>" | ">" ~/ Content ~/ ETag ) ) // FIXME tag must be balanced
+    val TagHeader = P( "<" ~ Name ~/ (WL ~ Attribute).rep ~ WL.? )
+    val ETag      = P( "</" ~ Name ~ WL.? ~ ">" )
+
+    val Attribute = P( Name ~ Eq ~/ AttValue )
+    val Eq        = P( WL.? ~ "=" ~ WL.? )
+    val AttValue  = P(
+      "\"" ~/ (CharQ | Reference).rep ~ "\"" |
+        "'" ~/ (CharA | Reference).rep ~ "'" |
+        ScalaExpr
+    )
+
+    val Content        = P( (CharData | Reference | ScalaExpr | XmlContent).rep )
+    val XmlContent: P0 = P( Unparsed | CDSect | PI | Comment | Element )
+
+    val ScalaExpr = P( "{" ~ Block ~ "}" )
+
+    val Unparsed = P( UnpStart ~/ UnpData ~ UnpEnd )
+    val UnpStart = P( "<xml:unparsed" ~/ (WL ~ Attribute).rep ~ WL.? ~ ">" )
+    val UnpEnd   = P( "</xml:unparsed>" )
+    val UnpData  = P( (!UnpEnd ~ AnyChar).rep )
+
+    val CDSect  = P( CDStart ~/ CData ~ CDEnd )
+    val CDStart = P( "<![CDATA[" )
+    val CData   = P( (!"]]>" ~ Char).rep )
+    val CDEnd   = P( "]]>" )
+
+    val Comment = P( "<!--" ~/ ComText ~ "-->" )
+    val ComText = P( (!"--" ~ Char).rep ~ ("-" ~ &("--")).? )
+
+    val PI         = P( "<?" ~ PITarget ~ PIProcText.? ~ "?>" )
+    val PITarget   = P( !(("X" | "x") ~ ("M" | "m") ~ ("L" | "l")) ~ Name )
+    val PIProcText = P( WL ~ (!"?>" ~ Char).rep )
+
+    val Reference = P( EntityRef | CharRef )
+    val EntityRef = P( "&" ~ Name ~/ ";" )
+    val CharRef   = P( "&#" ~ Num ~/ ";" | "&#x" ~ HexNum ~/ ";" )
+    val Num       = P( CharIn('0' to '9').rep )
+    val HexNum    = P( CharIn('0' to '9', 'a' to 'f', 'A' to 'F').rep )
+
+    val CharData = P( (!"{" ~ Char1 | "{{").rep(1) )
+
+    val Char   = P( AnyChar )
+    val Char1  = P( !("<" | "&") ~ Char )
+    val CharQ  = P( !"\"" ~ Char1 )
+    val CharA  = P( !"'" ~ Char1 )
+
+    val Name      = P( NameStart ~ NameChar.rep )
+    val NameStart = P( CharPred(isNameStart) ).opaque("NameStart")
+    val NameChar  = P( CharPred(isNameChar) ).opaque("NameChar")
+
+    val ElemPattern: P0 = P( TagPHeader ~/ ("/>" | ">" ~/ ContentP ~/ ETag ) )
+    val TagPHeader      = P( "<" ~ Name ~ WL.?  )
+
+    val ContentP: P0  = P( ( CharDataP | ScalaPatterns | ElemPattern ).rep )
+    val ScalaPatterns = P( "{" ~ Patterns ~ "}" )
+    val CharDataP     = P( "&" ~ CharData.? | CharData ) // matches weirdness of scalac parser on xml reference.
+
+    //================================================================================
+    // From `scala.xml.parsing.TokenTests`
```

Comment:
Most of LegacyScanner is copy-pasted from scala/scala, this is totally fine since it's so small compared to the total size of scala-xml. We try to avoid dependencies if possible, fastparse is an exception.